### ISSUE TITLE
False positives when functional components are rendered in StrictMode

### ISSUE THREAD

**Issue by justingrant** [welldone-software/why-did-you-render#78]
I'm seeing function components report false positives in StrictMode. I'm working on a repro case but I also think I've found the root cause in WDYR source, so filing this issue now. I'll add a repro if I can get it to repro reliably outside my (large) app.  it's also possible that this may be the root cause behind #16 and/or #3.  I wasn't sure though, so I'm opening a new issue. Feel free to resolve this as a duplicate if you want me to merge this info into another issue. 

The problem seems to be an assumption (excerpt below) that only class components are double-rendered in StrictMode but not function components.
https://github.com/welldone-software/why-did-you-render/blob/aeafe38c34eef6913c4cdf23e6bf20acee3a9166/src/utils.js#L6-L17

But looking at React's source, it looks like StrictMode double-rendering also happens with function components, but only if hooks are used.

https://github.com/facebook/react/blob/85d9655d6f72fc34b71b6b25b1f76a307db81c96/packages/react-reconciler/src/ReactFiberBeginWork.js#L635-L672

```js
  if (__DEV__) {
    ReactCurrentOwner.current = workInProgress;
    setCurrentPhase('render');
    nextChildren = renderWithHooks(
      current,
      workInProgress,
      Component,
      nextProps,
      context,
      renderExpirationTime,
    );
    if (
      debugRenderPhaseSideEffectsForStrictMode &&
      workInProgress.mode & StrictMode
    ) {
      // Only double-render components with Hooks
      if (workInProgress.memoizedState !== null) {
        nextChildren = renderWithHooks(
          current,
          workInProgress,
          Component,
          nextProps,
          context,
          renderExpirationTime,
        );
      }
    }
    setCurrentPhase(null);
  } else {
    nextChildren = renderWithHooks(
      current,
      workInProgress,
      Component,
      nextProps,
      context,
      renderExpirationTime,
    );
  }
```

BTW, the "ignore every other render" solution that's currently used for class components probably won't work as-is for function components, because the current StrictMode-handling solution is that it relies on components always being rendered twice. I'm not sure if this is always true, especially with function components. It's definitely not true when using React DevTools (see #77) because devtools will re-render function components [only once](https://github.com/facebook/react/issues/17670#issuecomment-568140251) in order to inspect hooks.

**Comment by vzaidman** [comment]
@justingrant 
hey!
I moved all strict mode tests to a separate file
plus added strict mode tests to functional components with hooks.

all seems to work as expected for now.

can you check if this still happens to you on the latest version of the library?

**Comment by vzaidman** [comment]
@justingrant

**Comment by justingrant** [comment]
Hi @vzaidman - sorry for delayed response, as I've been out for the holidays and am just catching up now.  Thanks for following up.

With the latest WDYR I'm not seeing the same false positives I was seeing before, which is great. Thanks for your fix!

I did previously change some of my code from useMemo to `useRef` to avoid this issue and #77.  I'm planning to change that code back to useMemo.  When I do that if I run into any false positives I'll file a new issue. Sound good?

Thanks again for your work building WDYR-- it's a really useful tool that's already saved me many hours.
